proposals/2019-02-12-design-for-a-python-toolchain.md

```diff
@@ -0,0 +1,246 @@
+
+## Example
+
+Here is a minimal example that defines a platform whose Python interpreters are located under a non-standard path. The example also defines a Python toolchain to accompany this platform.
+
+```python
+# //platform_defs:BUILD
+
+load("@bazel_tools//tools/python:toolchain.bzl", "py_runtime_pair")
+
+# Constraint values that represent that the system's "python2" and "python3"
+# executables are located under /usr/weirdpath.
+
+constraint_value(
+    name = "usr_weirdpath_python2",
+    constraint_setting = "@bazel_tools//tools/python:py2_interpreter_path",
+)
+
+constraint_value(
+    name = "usr_weirdpath_python3",
+    constraint_setting = "@bazel_tools//tools/python:py3_interpreter_path",
+)
+
+# A definition of a platform whose Python interpreters are under these paths.
+
+platform(
+    name = "my_platform",
+    constraint_values = [
+        ":usr_weirdpath_python2",
+        ":usr_weirdpath_python3",
+    ],
+)
+
+# Python runtime definitions that reify these system paths as BUILD targets.
+
+py_runtime(
+    name = "my_platform_py2_runtime",
+    interpreter_path = "/usr/weirdpath/python2",
+)
+
+py_runtime(
+    name = "my_platform_py3_runtime",
+    interpreter_path = "/usr/weirdpath/python3",
+)
+
+py_runtime_pair(
+    name = "my_platform_runtimes",
+    py2_runtime = ":my_platform_py2_runtime",
+    py3_runtime = ":my_platform_py3_runtime",
+)
+
+# A toolchain definition to expose these runtimes to toolchain resolution.
+
+toolchain(
+    name = "my_platform_python_toolchain",
+    # Since the Python interpreter is invoked at runtime on the target
+    # platform, there's no need to specify execution platform constraints here.
+    target_compatible_with = [
+        # Make sure this toolchain is only selected for a target platform that
+        # advertises that it has interpreters available under /usr/weirdpath.
+        ":usr_weirdpath_python2",
+        ":usr_weirdpath_python3",
+    ],
+    toolchain = ":my_platform_runtimes",
+    toolchain_type = "@bazel_tools//tools/python:toolchain_type",
+)
+```
+
+```python
+# //pkg:BUILD
+
+# An ordinary Python target to build.
+py_binary(
+    name = "my_pybin",
+    srcs = ["my_pybin.py"],
+    python_version = "PY3",
+)
+```
+
+```python
+# WORKSPACE
+
+# Register the custom Python toolchain so it can be chosen for my_platform.
+register_toolchains(
+    "//platform_defs:my_platform_python_toolchain",
+)
+```
+
+We can then build with
+
+```
+bazel build //pkg:my_pybin --platforms=//platform_defs:my_platform
+```
+
+and thanks to toolchain resolution, the resulting executable will automatically know to use the interpreter located at `/usr/weirdpath/python3`.
+
+If we had not defined a custom toolchain, then we'd be stuck with `autodetecting_python_toolchain`, which would fail at execution time if `/usr/weirdpath` were not on `PATH`. (It would also be slightly slower since it requires an extra invocation of the interpreter at execution time to confirm its version.)
```

Comment:
It's a per-target penalty, and it's paid at execution time, not build time. So each time you run a `py_binary` artifact (the stub script) built using this autodetecting toolchain, it will pay the cost of `exec`-ing to the extra wrapper script (as opposed to directly to the interpreter), searching directories on PATH for the system interpreter, and verifying its version by invoking it with the `-V` flag.

Fortunately Python does in fact have a `-V` flag, so we don't need to actually start up the true interpreter just to print `sys.version`. I didn't realize that when I wrote this parenthetical.

Originally I also had a `host_python_toolchain` in this proposal, that would autodetect the Python binary just once at WORKSPACE evaluation time, but only for the host platform. I removed it at katre@'s suggestion since it didn't seem to add much value.

In any case, users *should* be defining explicit toolchains for their platforms, which will avoid any inefficiency implied by this default fallback.